Blog post on slurm job builds (#14)
easybuilders/easybuild-blog: docs/posts/2026/04/jobs.md

---
authors:
  - micket
date: 2026-04-26
slug: jobs
hide:
  - navigation
---

# Building software with slurm jobs using EasyBuild

This blogpost serves as inspiration and get you started running jobs based on some of the stuff I run on my clusters.

Do you want to conveniently build modules in a hurry? Then you will love `--job`  combined with `-r`.

> 

Reviewer: ```suggestion

<!-- more -->

```

Or somewhere else near the top. The text above this will appear on the homepage of the blog. All of the text will appear on the individual blog post.

If you have multiple cluster or architectures, you may wish to place them, remember that you can set this via environment variables as well, e.g:
```bash
export EASYBUILD_JOB_OUTPUT_DIR=$HOME/log_${CLUSTER_NAME}_${ARCH}/
```

## Debug failed builds

If you have a problem build and the job logs isn't that useful, you can redirect tmpdir to persistent storage:
```
--tmpdir=/your/centrestorage/path/eb-tmp
```

## Monitoring the queue

### Simple
A handy alias to nicely show your queue can be made to keep track:

```bash
alias q='squeue --me -O jobid:10,tres-per-node:18,name:60,TimeUsed:10,reasonlist'
alias wq='watch -c \"squeue --me -O jobid:10,tres-per-node:18,name:60,TimeUsed:10,reasonlist\"'
```

### Colorful

Using `bat`, we can also color the output nicely

```bash
alias wq="watch -c \"squeue -u c3-builder -O jobid:10,tres-per-node:18,name:60,TimeUsed:10,reasonlist | sed 's/^ *//g' | sed 's/ \+/,/g' | bat -f -l csv --style plain --theme=ansi | column -s, -t | bat --style grid\""
```

### Advanced

Why not look at sacct as well? And color based on state?

> ```python=

Reviewer: ```suggestion
```python
```

Author: :( no line numbers...

Reviewer: Add the following to the `markdown_extensions` section in https://github.com/easybuilders/easybuild-blog/blob/main/mkdocs.yml#L11
```
  - pymdownx.highlight:
      anchor_linenums: true
  - pymdownx.superfences
```
Then change this to be
```
```python linenums="1"
```

    table = Table(show_header=True, header_style="bold", box=None)
    table.add_column("JOBID", justify="right")
    table.add_column("STATE")
    table.add_column("ARCH")
    table.add_column("GPU")
    table.add_column("TIME", justify="right")
    table.add_column("NODE")
    table.add_column("NAME")
    table.add_column("OUT")

    for j in jobs:
        state = j['state']
        stdout = ''
        if state.upper() in ("FAILED", "TIMEOUT"):
            state = f"[red]{state}[/red]"
            stdout = fix_path(j['stdout'], j['jobid'])
        elif state.upper() == "COMPLETED":
            state = f"[green]{state}[/green]"
        elif state.upper() == "CANCELLED":
            state = f"[yellow]{state}[/yellow]"
        elif state.upper() == "PENDING":
            state = f"[#888888]{state}[/#888888]"

        arch = j['arch']
        if arch == 'ZEN':
            arch = f"[red]{arch}[/red]"
        else:
            arch = f"[blue]{arch}[/blue]"

        table.add_row(j['jobid'], state, arch, j['gpu'], j['time'], j['node'], j['name'], stdout)
    console.print(table)

if __name__ == "__main__":
    main()
```

example running with `watch --color q_advanced`:

{{
asciinema(

>   '../../watch-q.cast',

Reviewer: ```suggestion
  '../../jobs.cast',
```

---
authors:
  - micket
date: 2026-04-26
slug: jobs
hide:
  - navigation
---

# Building software with slurm jobs using EasyBuild

This blogpost serves as inspiration and get you started running jobs based on some of the stuff I run on my clusters.

Do you want to conveniently build modules in a hurry? Then you will love `--job`  combined with `-r`.

<!-- more -->

## What is does

By specifying `eb -r --job Foobar-1.2.3-foss-2025b.eb` easybuild will
1. Resolves all dependencies
2. Pre-fetch all the sources
3. Submit jobs for each easyconfig that needs to be built
   - Jobs have dependency tracking based on their dependencies

And then you'll see
```
$ squeue --me

             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)
           9015294      vera Foobar-1      you PD       0:00      1 (Dependency)
           9015292      vera Stuff-1.      you  R       0:06      1 node-01
           9015293      vera Thing-2.      you  R       0:06      1 node-02
```

You can specify `--job-cores=X` to pick the size of your job.

## Passing SLURM flags

What if your job needs a GPU partition, or specify a TRES?

> Set them via environment variables

Reviewer 2: ```suggestion
Set them via environment variables:
```

```bash
SBATCH_GPUS_PER_NODE=H100:1 eb -r --job-cores=8 Foobar-1.2.3-foss-2025b-CUDA-12.8.0.eb
```

In fact, why not make an alias for it?

```bash
alias buildH100='SBATCH_GPUS_PER_NODE=H100:1 eb -r --job'
```

Depending on your cluster, you may find the following environment variables useful:

```bash
SBATCH_GPUS_PER_NODE
SBATCH_CONSTRAINT
SBATCH_PARTITION
SBATCH_ACCOUNT
```

but of course, many more exist.

## Collect build logs

If you don't want job logs in your current working directory when you submit, collect them to one place using `--job-output=/path/to/logs` .

> If you have multiple cluster or architectures, you may wish to place them, remember that you can set this via environment variables as well, e.g:

Reviewer 2: ```suggestion
If you have multiple cluster or architectures, you may wish to place them in specific locations.
Remember that you can set this via environment variables as well, e.g:
```

```bash
export EASYBUILD_JOB_OUTPUT_DIR=$HOME/log_${CLUSTER_NAME}_${ARCH}/
```

## Debug failed builds

If you have a problem build and the job logs isn't that useful, you can redirect tmpdir to persistent storage:
```bash
--tmpdir=/your/centrestorage/path/eb-tmp
```

## Monitoring the queue

### Simple

> A handy alias to nicely show your queue can be made to keep track:

Reviewer 2: ```suggestion
A handy alias to nicely keep track of your job queue can be made:
```

```bash
alias q='squeue --me -O jobid:10,tres-per-node:18,name:60,TimeUsed:10,reasonlist'
alias wq='watch -c \"squeue --me -O jobid:10,tres-per-node:18,name:60,TimeUsed:10,reasonlist\"'
```

### Colorful

> Using `bat`, we can also color the output nicely

Reviewer 2: ```suggestion
Using `bat`, we can even color the output nicely:
```

---
authors:
  - micket
date: 2026-04-26
slug: jobs
hide:
  - navigation
---

# Building software with slurm jobs using EasyBuild

This blogpost serves as inspiration and get you started running jobs based on some of the stuff I run on my clusters.

Do you want to conveniently build modules in a hurry? Then you will love `--job`  combined with `-r`.

<!-- more -->

## What is does

By specifying `eb -r --job Foobar-1.2.3-foss-2025b.eb` easybuild will
1. Resolves all dependencies
2. Pre-fetch all the sources
3. Submit jobs for each easyconfig that needs to be built
   - Jobs have dependency tracking based on their dependencies

And then you'll see
```
$ squeue --me

             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)
           9015294      vera Foobar-1      you PD       0:00      1 (Dependency)
           9015292      vera Stuff-1.      you  R       0:06      1 node-01
           9015293      vera Thing-2.      you  R       0:06      1 node-02
```

You can specify `--job-cores=X` to pick the size of your job.

## Passing SLURM flags

What if your job needs a GPU partition, or specify a TRES?

> Set them via environment variables:

Reviewer 2: ```suggestion
Set them via environment variables:

```

```bash
SBATCH_GPUS_PER_NODE=H100:1 eb -r --job-cores=8 Foobar-1.2.3-foss-2025b-CUDA-12.8.0.eb
```

In fact, why not make an alias for it?

```bash
alias buildH100='SBATCH_GPUS_PER_NODE=H100:1 eb -r --job'
```

Depending on your cluster, you may find the following environment variables useful:

```bash
SBATCH_GPUS_PER_NODE
SBATCH_CONSTRAINT
SBATCH_PARTITION
SBATCH_ACCOUNT
```

but of course, many more exist.

## Collect build logs

If you don't want job logs in your current working directory when you submit, collect them to one place using `--job-output=/path/to/logs` .

> If you have multiple cluster or architectures, you may wish to place them in specific locations.

Reviewer 2: ```suggestion
If you have multiple cluster or architectures, you may wish to place them in specific locations.

```

> Remember that you can set this via environment variables as well, e.g:

Reviewer 2: ```suggestion
Remember that you can set this via environment variables as well, e.g:

```

```bash
export EASYBUILD_JOB_OUTPUT_DIR=$HOME/log_${CLUSTER_NAME}_${ARCH}/
```

## Debug failed builds

If you have a problem build and the job logs isn't that useful, you can redirect tmpdir to persistent storage:
```bash
--tmpdir=/your/centrestorage/path/eb-tmp
```

## Monitoring the queue

### Simple

> A handy alias to nicely keep track of your job queue can be made:

Reviewer 2: ```suggestion
A handy alias to nicely keep track of your job queue can be made:

```

```bash
alias q='squeue --me -O jobid:10,tres-per-node:18,name:60,TimeUsed:10,reasonlist'
alias wq='watch -c \"squeue --me -O jobid:10,tres-per-node:18,name:60,TimeUsed:10,reasonlist\"'
```

### Colorful

> Using `bat`, we can even color the output nicely:

Reviewer 2: ```suggestion
Using `bat`, we can even color the output nicely:

```
